File: push-message/client_flutter/lib/socket_proxy.dart
import 'package:ezyfox_server_flutter_client/ezy_client.dart';
import 'package:ezyfox_server_flutter_client/ezy_clients.dart';
import 'package:ezyfox_server_flutter_client/ezy_config.dart';
import 'package:ezyfox_server_flutter_client/ezy_constants.dart';
import 'package:ezyfox_server_flutter_client/ezy_entities.dart';
import 'package:ezyfox_server_flutter_client/ezy_handlers.dart';

const ZONE_NAME = "push-message";
const APP_NAME = "push-message";

class SocketProxy {
  bool settedUp = false;
  late String username;
  late String password;
  late EzyClient _client;
  late Function(Map)? _messageCallback;
  late Function? _loggedInCallback;
  late Function? _disconnectedCallback;
  static final SocketProxy _INSTANCE = SocketProxy._();

  SocketProxy._() {}

  static SocketProxy getInstance() {
    return _INSTANCE;
  }

  void _setup() {
    EzyConfig config = EzyConfig();
    config.clientName = ZONE_NAME;
    EzyClients clients = EzyClients.getInstance();
    _client = clients.newDefaultClient(config);
    _client.setup.addEventHandler(EzyEventType.CONNECTION_FAILURE, _ConnectionFailureHandler(_disconnectedCallback!));
    _client.setup.addDataHandler(EzyCommand.HANDSHAKE, _HandshakeHandler());
    _client.setup.addDataHandler(EzyCommand.LOGIN, _LoginSuccessHandler());
    _client.setup.addDataHandler(EzyCommand.APP_ACCESS, _AppAccessHandler(_loggedInCallback!));
    var appSetup = _client.setup.setupApp(APP_NAME);
    appSetup.addDataHandler("message", _MessageResponseHandler((data) => {
      _messageCallback!(data)
    }));

  }

  void connectToServer(String username, String password) {
    if(!settedUp) {
      settedUp = true;
      _setup();
    }
    this.username = username;
    this.password = password;
    this._client.connect("localhost", 3005);
  }

  void onLoggedIn(Function callback) {
    this._loggedInCallback = callback;
  }

  void onDisconnected(Function callback) {
    this._disconnectedCallback = callback;
  }

  void onMessage(Function(Map) callback) {
    this._messageCallback = callback;
  }
}

class _ConnectionFailureHandler extends EzyConnectionFailureHandler {
  late Function _callback;

  _ConnectionFailureHandler(Function callback) {
    _callback = callback;
  }

  @override
  void onConnectionFailed(Map event) {
    _callback();
  }
}

class _HandshakeHandler extends EzyHandshakeHandler {
  @override
  List getLoginRequest() {
    var request = [];
    request.add(ZONE_NAME);
    request.add(SocketProxy.getInstance().username);
    request.add(SocketProxy.getInstance().password);
    request.add([]);
    return request;
  }
}

class _LoginSuccessHandler extends EzyLoginSuccessHandler {
  @override
  void handleLoginSuccess(responseData) {
    client.send(EzyCommand.APP_ACCESS, [APP_NAME]);
  }
}

class _AppAccessHandler extends EzyAppAccessHandler {
  late Function _callback;

  _AppAccessHandler(Function callback) {
    _callback = callback;
  }

  @override
  void postHandle(EzyApp app, List data) {
    _callback();
  }
}

class _MessageResponseHandler extends EzyAppDataHandler<Map> {

  late Function(Map) _callback;

  _MessageResponseHandler(Function(Map) callback) {
    _callback = callback;
  }

  @override
  void handle(EzyApp app, Map data) {
    _callback(data);
  }
}

File: push-message/client-react-native/socket/SocketProxy.js
import Ezy from 'ezyfox-server-react-native-client';
import Mvc from 'mvc-es6';
import {ZONE_NAME, APP_NAME, Command} from './SocketConstants';

class SocketProxy {
  static instance = null;

  static getInstance() {
    if (!SocketProxy.instance) {
      SocketProxy.instance = new SocketProxy();
    }
    return SocketProxy.instance;
  }

  setup() {
    let config = new Ezy.ClientConfig();
    config.clientName = ZONE_NAME;
    config.enableSSL = true;
    config.enableDebug = true;
    let clients = Ezy.Clients.getInstance();
    clients.newDefaultClient(config, client => {
      this.doSetup(client);
    });
  }

  doSetup(client) {
    let mvc = Mvc.getInstance();
    let models = mvc.models;
    let handshakeHandler = new Ezy.HandshakeHandler();
    handshakeHandler.getLoginRequest = function () {
      let connection = models.connection;
      let username = connection.username;
      let password = connection.password;
      return [ZONE_NAME, username, password, []];
    };

    let loginSuccessHandler = new Ezy.LoginSuccessHandler();
    loginSuccessHandler.handleLoginSuccess = function () {
      let accessAppRequest = [APP_NAME, []];
      this.client.send(Ezy.Command.APP_ACCESS, accessAppRequest);
    };

    let loginErrorHandler = new Ezy.LoginErrorHandler();
    loginErrorHandler.handleLoginError = function (event) {
      let loginController = mvc.getController('login');
      loginController.updateViews('loginError', event[1]);
    };

    let accessAppHandler = new Ezy.AppAccessHandler();
    accessAppHandler.postHandle = function (app, data) {
      let routerController = mvc.getController('router');
      routerController.updateViews('change', 'message');
    };

    let exitAppHandler = new Ezy.AppExitHandler();
    exitAppHandler.postHandle = function (app, data) {
      this.client.close();
    };

    let connectionSuccessHandler = new Ezy.ConnectionSuccessHandler();
    connectionSuccessHandler.postHandle = function (event) {
      console.log('connected to server');
    };

    let disconnectionHandler = new Ezy.DisconnectionHandler();
    disconnectionHandler.preHandle = function (event) {
      let routerController = mvc.getController('router');
      routerController.updateViews('change', 'login');
    };

    let setup = client.setup;
    setup.addEventHandler(
      Ezy.EventType.CONNECTION_SUCCESS,
      connectionSuccessHandler,
    );
    setup.addEventHandler(Ezy.EventType.DISCONNECTION, disconnectionHandler);
    setup.addDataHandler(Ezy.Command.HANDSHAKE, handshakeHandler);
    setup.addDataHandler(Ezy.Command.LOGIN, loginSuccessHandler);
    setup.addDataHandler(Ezy.Command.LOGIN_ERROR, loginErrorHandler);
    setup.addDataHandler(Ezy.Command.APP_ACCESS, accessAppHandler);
    setup.addDataHandler(Ezy.Command.APP_EXIT, exitAppHandler);
    let setupApp = setup.setupApp(APP_NAME);

    let messageController = mvc.getController('message');

    setupApp.addDataHandler(Command.MESSAGE, function (app, data) {
      messageController.updateViews(Command.MESSAGE, data);

    });

    return client;
  }

  connect() {
    let client = this.getClient();
    client.connect('127.0.0.1', 3005);
  }

  isConnected() {
    let client = this.getClient();
    let connected = client && client.isConnected();
    return connected;
  }

  getClient() {
    let clients = Ezy.Clients.getInstance();
    let client = clients.getDefaultClient();
    return client;
  }
}

export default SocketProxy;